lib/mailbox-handler.js

```diff
@@ -19,99 +20,92 @@ class MailboxHandler {
+
+        this.settingsHandler = new SettingsHandler({ db: this.database });
     }
 
     create(user, path, opts, callback) {
-        this.database.collection('mailboxes').findOne(
-            {
-                user,
-                path
-            },
-            (err, mailboxData) => {
-                if (err) {
-                    return callback(err);
-                }
-                if (mailboxData) {
-                    const err = new Error('Mailbox creation failed with code MailboxAlreadyExists');
-                    err.code = 'ALREADYEXISTS';
-                    err.responseCode = 400;
-                    return callback(err, 'ALREADYEXISTS');
-                }
+        this.createAsync(user, path, opts)
+            .then(mailboxData => callback(null, ...Object.values(mailboxData)))
+            .catch(err => callback(err));
+    }
 
-                this.users.collection('users').findOne(
-                    {
-                        _id: user
-                    },
-                    {
-                        projection: {
-                            retention: true
-                        }
-                    },
-                    (err, userData) => {
-                        if (err) {
-                            return callback(err);
-                        }
+    async createAsync(user, path, opts) {
+        let mailboxData = await this.database.collection('mailboxes').findOne({ user, path });
 
-                        if (!userData) {
-                            const err = new Error('This user does not exist');
-                            err.code = 'UserNotFound';
-                            err.responseCode = 404;
-                            return callback(err, 'UserNotFound');
-                        }
+        if (mailboxData) {
+            const err = new Error('Mailbox creation failed with code MailboxAlreadyExists');
+            err.code = 'ALREADYEXISTS';
+            err.responseCode = 400;
+            throw err;
+        }
 
-                        mailboxData = {
-                            _id: new ObjectId(),
-                            user,
-                            path,
-                            uidValidity: Math.floor(Date.now() / 1000),
-                            uidNext: 1,
-                            modifyIndex: 0,
-                            subscribed: true,
-                            flags: [],
-                            retention: userData.retention
-                        };
+        const mailboxCountForUser = await this.database.collection('mailboxes').countDocuments({ user });
 
-                        Object.keys(opts || {}).forEach(key => {
-                            if (!['_id', 'user', 'path'].includes(key)) {
-                                mailboxData[key] = opts[key];
-                            }
-                        });
+        if (mailboxCountForUser > (await this.settingsHandler.get('const:max:mailboxes', {}))) {
```

Comment:
The second argument (options object) is optional so that the empty object can be skipped. I have no idea why it is used in other places. Maybe there were some options set initially and later got removed, but the empty options object remained in place 🤔

Reply:
Removed unnecessary second parram. Indeed it is not required and everything seems to work just fine without it